### PR TITLE
Parallelize batch processing in E2E scripts

### PR DESCRIPTION
**Pull request by nlordell**
This PR uses the `rayon` library to parallelize processing each batch to speed things up. This takes advantage of the fact that the auction state and `pricegraph` can be computed "on-the fly" from an immutable event collection in order to take full advantage of all CPU cores :rocket: 

Additionally, the nasty threading bits and dealing with synchronization has been abstracted behind a `Reporting` trait to make things easier and reduce boiler plate.

:warning: There is one disadvantage in that the output is **no longer in order**. I don't think that is a huge issue, since the data can be ordered when it is being processed and analyzed.

### Test Plan

Before `233.85s` :turtle: 
```
$ time -p target/release/historic_prices --orderbook-file target/orderbook.mainnet.bin
54678 / 54678 [==============================================] 100.00 % 230.39/s
Processed 13743 token prices: bad estimates 45.19%, average error 27643.61%.
real 237.53
user 233.85
sys 2.78
```

After `53.27s` :rocket: 
```
$ time -p target/release/historic_prices --orderbook-file target/orderbook.mainnet.bin
54678 / 54678 [=============================================] 100.00 % 1033.13/s
Processed 13743 token prices: bad estimates 45.22%, average error 27938.76%.
real 53.27
user 397.93
sys 2.05
```

A 4.5x increase on an 8-core machine. Notice how the `user` time is almost double. We are losing a lot of time dealing with thread synchronization, and might be able to speed it up even a bit more.

### CODE REVIEW

File: e2e/src/bin/historic_trades.rs [hunk 1]
```diff
@@ -34,7 +34,7 @@ fn main() -> Result<()> {
     let mut report = Report::new(File::create(options.output_dir.join("trades.csv"))?);
     report.header()?;
 
-    cmd::for_each_batch(&options.orderbook_file, |history, batch| {
+    cmd::for_each_batch_sync(&options.orderbook_file, |history, batch| {
```
Comment (nlordell) [inline e2e/src/bin/historic_trades.rs:37]: I will migrate this script in a follow up PR.

File: e2e/src/cmd.rs [hunk 2]
```diff
@@ -22,3 +24,101 @@ pub fn for_each_batch(
...
+                .try_for_each_with(sampler.clone(), move |sampler, batch| {
+                    handler(&sampler, &history, batch)?;
+                    sampler.batch_complete()
+                })
+            {
+                sampler.error(err);
+            }
+        });
+
+        while let Ok(result) = samples_rx.recv() {
```
Comment (e00E) [inline e2e/src/cmd.rs:60]: This should probably be in a `scope.spawn` too because the intention is to have this run in parallel.

**Review by e00E — APPROVED** [review]

File: e2e/src/cmd.rs (hunk 2 above)
Comment (nlordell) [reply e2e/src/cmd.rs:60]: So, the idea here was to have the main thread update the progress bar iterator and record the samples while a background thread processes the parallel . I chose this for a couple of reasons:
1. Not require another thread for this, we can just use the main thread and then join once we finish collecting
2. Not require `report` to also be `Send`

File: e2e/src/cmd.rs [hunk 3]
```diff
@@ -22,3 +24,101 @@ pub fn for_each_batch(
...
+
+/// A trait for reporting samples collected from each batch run.
+pub trait Reporting {
+    /// The sample type.
+    type Sample: Send;
+    /// The summary type.
+    type Summary;
+
+    /// Record a sample.
+    fn sample(&mut self, sample: Self::Sample) -> Result<()>;
```
Comment (fleupold) [inline e2e/src/cmd.rs:120]: ubernanonit: _to sample_ for me is to select a record from a set of candidates not to actually so maybe record_sample would be a better verb here?

**Review by fleupold — APPROVED** [review]
Very nice abstraction and impressive speedup

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia0.giphy.com%2Fmedia%2F26uf9QPzzlKPvQG5O%2Fgiphy.gif&token=warpspeed" data-gitmeme-token="warpspeed"></a>

File: e2e/src/cmd.rs (hunk 3 above)
Comment (nlordell) [reply e2e/src/cmd.rs:120]: Makes more sense that way. To "sample" is more about reading and less about recording as you pointed it.

**Pull request merged**